### PR TITLE
fix(coderd): subscribe to workspace when streaming agent logs to detect outdated build

### LINKED ISSUE #9721: test flake: TestWorkspaceAgentStartupLogs/Close_logs_on_outdated_build
TestWorkspaceAgentStartupLogs/Close_logs_on_outdated_build

```
    workspaceagents_test.go:419: 
        	Error Trace:	C:/a/coder/coder/coderd/workspaceagents_test.go:419
        	Error:      	context done while waiting for logs close
        	Test:       	TestWorkspaceAgentStartupLogs/Close_logs_on_outdated_build
```

https://github.com/coder/coder/actions/runs/6219337218/job/16877236793?pr=9717

### PR DESCRIPTION
**Pull request by mafredri**
Because of lax pubsub propagation (e.g. agents channels aren't signaled that they're outdated), we were hitting the 1 minute timeout to detect that the workspace build was outdated leading to test flakes.

This changes improves outdated build detection by subscribing to workspace updates as well.

Fixes #9721

### CODE REVIEW

File: codersdk/workspaceagents.go [hunk 1]
```diff
@@ -703,7 +703,7 @@ func (c *Client) WorkspaceAgentLogsAfter(ctx context.Context, agentID uuid.UUID,
 		}
 		return nil, nil, ReadBodyAsError(res)
 	}
-	logChunks := make(chan []WorkspaceAgentLog)
+	logChunks := make(chan []WorkspaceAgentLog, 1)
```
Comment (mafredri) [inline codersdk/workspaceagents.go:706]: Review: Bonus change to speed up detection of closed connection.

File: coderd/workspaceagents.go [hunk 2]
```diff
@@ -570,11 +570,28 @@ func (api *API) workspaceAgentLogs(rw http.ResponseWriter, r *http.Request) {
...
 
+	workspaceNotifyCh := make(chan struct{}, 1)
 	notifyCh := make(chan struct{}, 1)
 	// Allow us to immediately check if we missed any logs
 	// between initial fetch and subscribe.
 	notifyCh <- struct{}{}
 
+	// Subscribe to workspace to detect new builds.
+	closeSubscribeWorkspace, err := api.Pubsub.Subscribe(codersdk.WorkspaceNotifyChannel(workspace.ID), func(_ context.Context, _ []byte) {
+		// The message is not important, we're tracking lastSentLogID manually.
```
Comment (mafredri) [inline coderd/workspaceagents.go:581]:
```suggestion
```

**Comment by spikecurtis** [comment]
I'm afraid I don't understand what you mean by "lax pubsub propagation," or what was wrong with the previous test that caused the failure.  While I agree it's an improvement to directly receive notification of workspace changes, I'm worried there is still a bug in log notifications.

**Comment by mafredri** [comment]
> I'm afraid I don't understand what you mean by "lax pubsub propagation,"

I was referring to how we separate different pars of a workspace into separate pubsub channels (in this case workspace and agent log updates), but you’re required to have the domain knowledge of how these channels are connected vs it being handled in places like workspace stop. The domain knowledge here being that there will be no more logs when the workspace is stopped. And what could instead happen ids that airspace stop also sends an eof on the agent log channels, hence lax.

> or what was wrong with the previous test that caused the failure.

The previous test has that second log push which, I’m guessing (since it’s impossible to confirm from the logs), raced with the newly started workspace build since the build wasn’t complete. The log push was intended to trigger the coderd loop to recheck if the agent is outdated which was the racy behavior.

> While I agree it's an improvement to directly receive notification of workspace changes, I'm worried there is still a bug in log notifications.

Are you thinking about this particular case that lead to the test flake or some other buggy behavior with the log notifications?

**Comment by spikecurtis** [comment]
> The previous test has that second log push which, I’m guessing (since it’s impossible to confirm from the logs), raced with the newly started workspace build (since we don’t wait for build completion).

I don't think this explanation is consistent with the logs from the failed test, which show that the build completed a couple hundred ms before the second log is PATCHed.

```
    t.go:85: 2023-09-18 07:15:32.500 [debu]  provisionerd.runner: sending CompletedJob  job_id=34bea8ee-c5a1-4b78-a78c-1cba64ecb3c4  template_name=stupefied-turing1-8pH  template_version=agitated_gould2  workspace_build_id=e000bb28-88be-4438-b78d-18ade7ede426  workspace_id=effb2a01-0e53-4ae2-95e9-5c85d72f9391  workspace_name=competent-wing4-wKu  workspace_owner=testuser  workspace_transition=start
    t.go:85: 2023-09-18 07:15:32.500 [info]  echo: recv done on Session  session_id=37bbc0bd-ce7c-481f-93dc-2fbe25f94c3f  error=EOF
    t.go:85: 2023-09-18 07:15:32.500 [debu]  inmem-provisionerd-serene_chatelet8: stage CompleteJob starting  job_id=34bea8ee-c5a1-4b78-a78c-1cba64ecb3c4
    t.go:85: 2023-09-18 07:15:32.501 [debu]  inmem-provisionerd-serene_chatelet8: triggering workspace notification after agent timeout  workspace_build_id=e000bb28-88be-4438-b78d-18ade7ede426  timeout=0s
    t.go:85: 2023-09-18 07:15:32.502 [debu]  inmem-provisionerd-serene_chatelet8: stage CompleteJob done  job_id=34bea8ee-c5a1-4b78-a78c-1cba64ecb3c4
    t.go:85: 2023-09-18 07:15:32.502 [debu]  provisionerd.runner: sent CompletedJob  job_id=34bea8ee-c5a1-4b78-a78c-1cba64ecb3c4  template_name=stupefied-turing1-8pH  template_version=agitated_gould2  workspace_build_id=e000bb28-88be-4438-b78d-18ade7ede426  workspace_id=effb2a01-0e53-4ae2-95e9-5c85d72f9391  workspace_name=competent-wing4-wKu  workspace_owner=testuser  workspace_transition=start
    t.go:85: 2023-09-18 07:15:32.502 [debu]  provisionerd: start of acquireAndRunOne
    t.go:85: 2023-09-18 07:15:32.502 [debu]  acquirer: acquiring job  worker_id=c11bf591-0daa-45f5-943a-daae2c371e99  provisioner_types="[echo terraform]"  tags={"scope":"organization"}
    t.go:85: 2023-09-18 07:15:32.502 [debu]  acquirer: got clearance to call database  worker_id=c11bf591-0daa-45f5-943a-daae2c371e99  provisioner_types="[echo terraform]"  tags={"scope":"organization"}
    t.go:85: 2023-09-18 07:15:32.502 [debu]  acquirer: no job available  worker_id=c11bf591-0daa-45f5-943a-daae2c371e99  provisioner_types="[echo terraform]"  tags={"scope":"organization"}
    t.go:85: 2023-09-18 07:15:32.505 [debu]  echo: cleaned up work directory  session_id=37bbc0bd-ce7c-481f-93dc-2fbe25f94c3f
    t.go:85: 2023-09-18 07:15:32.765 [debu]  PATCH  host=localhost:55352  path=/api/v2/workspaceagents/me/logs  proto=HTTP/1.1  remote_addr=127.0.0.1  took=0s  status_code=200  latency_ms=0  request_id=78d90611-9ebf-43eb-8eb2-8c14ebb8b7af
    workspaceagents_test.go:419: 
        	Error Trace:	C:/a/coder/coder/coderd/workspaceagents_test.go:419
        	Error:      	context done while waiting for logs close
        	Test:       	TestWorkspaceAgentStartupLogs/Close_logs_on_outdated_build
    t.go:85: 2023-09-18 07:15:51.396 [info]  provisionerd: attempting graceful shutdown
    t.go:85: 2023-09-18 07:15:51.396 [debu]  provisionerd: sending acquire job cancel
```

That's why I'm still suspicious about the logic that notifies about new logs and triggers updates on the channel

**Comment by mafredri** [comment]
@spikecurtis that’s what I meant with it not being verifiable in the logs. We log HTTP requests when they _end_, but not when they start.

There’s no telling whether or not it completed this instant or a second ago: https://github.com/coder/coder/blob/b358e3d5588b5a82e72457b52f315f2738c8744c/coderd/httpmw/logger.go#L62

Perhaps we should include start time as a kv as well.

**Comment by spikecurtis** [comment]
> There’s no telling whether or not it completed this instant or a second ago

We compute the latency, which, along with the end time should indicate the start time.

https://github.com/coder/coder/blob/b358e3d5588b5a82e72457b52f315f2738c8744c/coderd/httpmw/logger.go#L41

**Comment by mafredri** [comment]
> We compute the latency, which, along with the end time should indicate the start time.
> 
> https://github.com/coder/coder/blob/b358e3d5588b5a82e72457b52f315f2738c8744c/coderd/httpmw/logger.go#L41

We don’t log the end time either, and we don’t know when that log will be scheduled to run.

**Review by spikecurtis — APPROVED** [review]

**Pull request merged**